# Casos de covid en los estados unidos por condado

Se nos presentan datos de los muertos por covid en los Estados Unidos por Condado, debemos realizar un proceso de ETL en 3 archivos csv distintos.

## Importamos las librerías correspondientes

In [1]:
import pandas as pd
from numpy import nan
import sqlalchemy
import psycopg2

## 1. Extract

Realizamos la extracción de datos desde los archivos csv que se nos entregaron, los cuales por conveniencia estarán en el mismo directorio.
También deberemos analizar la composición de los datos obtenidos.

Notamos que los datos están separados por "/" en el caso de los condados, por "&" en el caso de los casos diarios y por "," en los estados. Son agregados a sus respectivos data frames

In [2]:
df_counties = pd.read_csv(filepath_or_buffer="counties_us.csv", sep="/", encoding='utf-8')
df_daily = pd.read_csv(filepath_or_buffer="daily_cases_us.csv", sep="&", encoding='utf-8')
df_states = pd.read_csv(filepath_or_buffer="states_us.csv", sep=",", encoding='utf-8')

Imprimimos cada uno de los data frames, para ver que columnas y tipos de datos los componen.

In [3]:
df_counties

fIps               cOUnty        statE statE_cOdE    malE  fEmalE  \
0      1001       AUtaUga CoUnty      Alabama         AL   26874   28326   
1      1003       BaldwIn CoUnty      Alabama         AL  101188  106919   
2      1005       BarboUr CoUnty      Alabama         AL   13697   12085   
3      1007          BIbb CoUnty      Alabama         AL   12152   10375   
4      1009        BloUnt CoUnty      Alabama         AL   28434   29211   
...     ...                  ...          ...        ...     ...     ...   
3215  72145  Vega Baja MUnIcIpIo  PuertO RicO        NaN   25580   27791   
3216  72147    VIeqUes MUnIcIpIo  PuertO RicO        NaN    4332    4439   
3217  72149   VIllalba MUnIcIpIo  PuertO RicO        NaN   11169   11824   
3218  72151    YabUcoa MUnIcIpIo  PuertO RicO        NaN   16541   17608   
3219  72153      YaUco MUnIcIpIo  PuertO RicO        NaN   17475   18964   

      mEdIan_agE        lat       lOng  
0           37.8  32.534923 -86.642730  
1           42.8  30.727479 -87.722564  
2           39.9  31.869581 -85.393210  
3           39.9  32.998628 -87.126475  
4           40.8  33.980869 -86.567380  
...          ...        ...        ...  
3215        40.7  18.428461 -66.397926  
3216        43.6  18.122662 -65.439095  
3217        38.8  18.128155 -66.472816  
3218        42.5  18.070468 -65.896311  
3219        43.0  18.079728 -66.858276  

[3220 rows x 9 columns]

In [4]:
df_daily

datE      cOUnty       statE     fIps  casEs  dEaths
0        2020!%01!%21   SnohomIsh  WashingtOn  53061.0      1     0.0
1        2020!%01!%22   SnohomIsh  WashingtOn  53061.0      1     0.0
2        2020!%01!%23   SnohomIsh  WashingtOn  53061.0      1     0.0
3        2020!%01!%24        Cook    IllinOis  17031.0      1     0.0
4        2020!%01!%24   SnohomIsh  WashingtOn  53061.0      1     0.0
...               ...         ...         ...      ...    ...     ...
2502827  2022!%05!%13  Sweetwater     WyOming  56037.0  11088   126.0
2502828  2022!%05!%13       Teton     WyOming  56039.0  10074    16.0
2502829  2022!%05!%13       UInta     WyOming  56041.0   5643    39.0
2502830  2022!%05!%13    WashakIe     WyOming  56043.0   2358    44.0
2502831  2022!%05!%13      Weston     WyOming  56045.0   1588    18.0

[2502832 rows x 6 columns]

In [5]:
df_states

cOUnty        statE statE_cOdE  pOpUlatIOn
0          Autauga County      Alabama         AL       55200
1          Baldwin County      Alabama         AL      208107
2          Barbour County      Alabama         AL       25782
3             Bibb County      Alabama         AL       22527
4           Blount County      Alabama         AL       57645
...                   ...          ...        ...         ...
3215  Vega Baja Municipio  Puerto Rico        NaN       53371
3216    Vieques Municipio  Puerto Rico        NaN        8771
3217   Villalba Municipio  Puerto Rico        NaN       22993
3218    Yabucoa Municipio  Puerto Rico        NaN       34149
3219      Yauco Municipio  Puerto Rico        NaN       36439

[3220 rows x 4 columns]

Notamos que las columnas están sucias y además hay datos faltantes o mal escritos, por lo que debemos pasar a un proceso de transformación.

## 2. Transform

### Separación de Datos

Cambiamos los nombres de las columnas, para poder indexar los datos correctamente.

In [6]:
colsCounty = ['fips', 'county', 'state', 'state_code', 'male', 'female', 'median_age', 'lat', 'long']
colsDays = ['date', 'county', 'state', 'fips', 'cases', 'deaths']
colsStates = ['county', 'state', 'state_code', 'population']
df_counties.columns = colsCounty
df_daily.columns = colsDays
df_states.columns = colsStates

Obtenemos los tipos de datos para cada data frame.

In [7]:
df_counties.dtypes #tipos de datos

fips            int64
county         object
state          object
state_code     object
male            int64
female          int64
median_age    float64
lat           float64
long          float64
dtype: object

In [8]:
df_daily.dtypes

date       object
county     object
state      object
fips      float64
cases       int64
deaths    float64
dtype: object

In [9]:
df_states.dtypes

county        object
state         object
state_code    object
population     int64
dtype: object

Obtenemos información de cada data frame, para ver cuantos datos nulos tenemos.

In [10]:
df_counties.info() #nos da información más detallada

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3220 entries, 0 to 3219
Data columns (total 9 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   fips        3220 non-null   int64  
 1   county      3220 non-null   object 
 2   state       3220 non-null   object 
 3   state_code  3141 non-null   object 
 4   male        3220 non-null   int64  
 5   female      3220 non-null   int64  
 6   median_age  3220 non-null   float64
 7   lat         3220 non-null   float64
 8   long        3220 non-null   float64
dtypes: float64(3), int64(3), object(3)
memory usage: 226.5+ KB


Notamos que para el caso de los condados, la columna State_code solo tiene 3141 datos no nulos.

In [11]:
df_counties["state_code"].unique() #vemos cuantos datos distintos tiene

array(['AL', 'AK', 'AZ', 'AR', 'CA', 'CO', 'CT', 'DE', nan, 'FL', 'GA',
       'HI', 'ID', 'IL', 'IN', 'IA', 'KS', 'KY', 'LA', 'ME', 'MD', 'MA',
       'MI', 'MN', 'MS', 'MO', 'MT', 'NE', 'NV', 'NH', 'NJ', 'NM', 'NY',
       'NC', 'ND', 'OH', 'OK', 'OR', 'PA', 'RI', 'SC', 'SD', 'TN', 'TX',
       'UT', 'VT', 'VA', 'WA', 'WV', 'WI', 'WY'], dtype=object)

In [12]:
df_counties.isnull().sum() #con este método podemos sumar los datos nulos por columna

fips           0
county         0
state          0
state_code    79
male           0
female         0
median_age     0
lat            0
long           0
dtype: int64

Volvemos a hacer el mismo procedimientos para los data frames de los casos diarios y los estados.

In [13]:
df_daily.info() #vemos información más detallada

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2502832 entries, 0 to 2502831
Data columns (total 6 columns):
 #   Column  Dtype  
---  ------  -----  
 0   date    object 
 1   county  object 
 2   state   object 
 3   fips    float64
 4   cases   int64  
 5   deaths  float64
dtypes: float64(2), int64(1), object(3)
memory usage: 114.6+ MB


In [14]:
df_daily.isnull().sum() #sumamos los nulos por columna

date          0
county        0
state         0
fips      23678
cases         0
deaths    57605
dtype: int64

In [15]:
df_states.info() #información del data frame

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3220 entries, 0 to 3219
Data columns (total 4 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   county      3220 non-null   object
 1   state       3220 non-null   object
 2   state_code  3141 non-null   object
 3   population  3220 non-null   int64 
dtypes: int64(1), object(3)
memory usage: 100.8+ KB


In [16]:
df_states['state_code'].unique()

array(['AL', 'AK', 'AZ', 'AR', 'CA', 'CO', 'CT', 'DE', nan, 'FL', 'GA',
       'HI', 'ID', 'IL', 'IN', 'IA', 'KS', 'KY', 'LA', 'ME', 'MD', 'MA',
       'MI', 'MN', 'MS', 'MO', 'MT', 'NE', 'NV', 'NH', 'NJ', 'NM', 'NY',
       'NC', 'ND', 'OH', 'OK', 'OR', 'PA', 'RI', 'SC', 'SD', 'TN', 'TX',
       'UT', 'VT', 'VA', 'WA', 'WV', 'WI', 'WY'], dtype=object)

In [17]:
df_states.isnull().sum() #obtenemos la suma de nulos por columna

county         0
state          0
state_code    79
population     0
dtype: int64

### Datos Nulos

Para el caso de los datos nulos tomamos la desición de eliminar los menos posibles, ya que mientras más datos se eliminan, más información perdemos.

Notamos que para el caso del data frame de los condados, la columna State_code retorna nulo en caso que el estado sea Puerto Rico o sea District of Columbia, son los 2 únicos casos, en base a eso podemos reemplazar dichos datos nulos por su correspondiente codigo estatal.

In [18]:
df_counties.loc[(df_counties['state_code'].isnull()) & (df_counties['state'] == 'PuertO RicO'), 'state_code'] = 'PR'            #la función loc nos permite agregar varias sentencias condicionales y reemplazar un valor por otro
df_counties.loc[(df_counties['state_code'].isnull()) & (df_counties['state'] == 'District Of COlumbia'), 'state_code'] = 'DC'
df_counties

fips               county        state state_code    male  female  \
0      1001       AUtaUga CoUnty      Alabama         AL   26874   28326   
1      1003       BaldwIn CoUnty      Alabama         AL  101188  106919   
2      1005       BarboUr CoUnty      Alabama         AL   13697   12085   
3      1007          BIbb CoUnty      Alabama         AL   12152   10375   
4      1009        BloUnt CoUnty      Alabama         AL   28434   29211   
...     ...                  ...          ...        ...     ...     ...   
3215  72145  Vega Baja MUnIcIpIo  PuertO RicO         PR   25580   27791   
3216  72147    VIeqUes MUnIcIpIo  PuertO RicO         PR    4332    4439   
3217  72149   VIllalba MUnIcIpIo  PuertO RicO         PR   11169   11824   
3218  72151    YabUcoa MUnIcIpIo  PuertO RicO         PR   16541   17608   
3219  72153      YaUco MUnIcIpIo  PuertO RicO         PR   17475   18964   

      median_age        lat       long  
0           37.8  32.534923 -86.642730  
1           42.8  30.727479 -87.722564  
2           39.9  31.869581 -85.393210  
3           39.9  32.998628 -87.126475  
4           40.8  33.980869 -86.567380  
...          ...        ...        ...  
3215        40.7  18.428461 -66.397926  
3216        43.6  18.122662 -65.439095  
3217        38.8  18.128155 -66.472816  
3218        42.5  18.070468 -65.896311  
3219        43.0  18.079728 -66.858276  

[3220 rows x 9 columns]

In [19]:
df_counties.isnull().sum() #verificamos que ya no queden datos nulos

fips          0
county        0
state         0
state_code    0
male          0
female        0
median_age    0
lat           0
long          0
dtype: int64

Podemos hacer algo similar para el caso del data frame de los estados.

In [20]:
df_states.loc[(df_states['state_code'].isnull()) & (df_states['state'] == 'Puerto Rico'), 'state_code'] = 'PR'          #Cambiamos los valores de los nan por el respectivo codigo estatal
df_states.loc[(df_states['state_code'].isnull()) & (df_states['state'] == 'District of Columbia'), 'state_code'] = 'DC'
df_states

county        state state_code  population
0          Autauga County      Alabama         AL       55200
1          Baldwin County      Alabama         AL      208107
2          Barbour County      Alabama         AL       25782
3             Bibb County      Alabama         AL       22527
4           Blount County      Alabama         AL       57645
...                   ...          ...        ...         ...
3215  Vega Baja Municipio  Puerto Rico         PR       53371
3216    Vieques Municipio  Puerto Rico         PR        8771
3217   Villalba Municipio  Puerto Rico         PR       22993
3218    Yabucoa Municipio  Puerto Rico         PR       34149
3219      Yauco Municipio  Puerto Rico         PR       36439

[3220 rows x 4 columns]

In [21]:
df_states.isnull().sum() #verificamos que ya no queden nulos

county        0
state         0
state_code    0
population    0
dtype: int64

Para el data frame de los casos diarios es un tema distinto.

Primero sabemos que hay 57605 valores nulos en las muertes, desde el punto de vista estadístico no sirve de mucho saber cuantos contagiados tuviste si no puedes saber cuantos de ellos murieron.
Es por ello que consideramos que es mejor eliminar dichas filas, además al ser más de 2 millones 500 mil datos en total, eliminar esos 57605 datos no afecta tanto al resultado final.

In [22]:
df_daily = df_daily.dropna(subset=['deaths']) #eliminamos la tupla cada vez que sea nulo en las muertes
df_daily

date      county       state     fips  cases  deaths
0        2020!%01!%21   SnohomIsh  WashingtOn  53061.0      1     0.0
1        2020!%01!%22   SnohomIsh  WashingtOn  53061.0      1     0.0
2        2020!%01!%23   SnohomIsh  WashingtOn  53061.0      1     0.0
3        2020!%01!%24        Cook    IllinOis  17031.0      1     0.0
4        2020!%01!%24   SnohomIsh  WashingtOn  53061.0      1     0.0
...               ...         ...         ...      ...    ...     ...
2502827  2022!%05!%13  Sweetwater     WyOming  56037.0  11088   126.0
2502828  2022!%05!%13       Teton     WyOming  56039.0  10074    16.0
2502829  2022!%05!%13       UInta     WyOming  56041.0   5643    39.0
2502830  2022!%05!%13    WashakIe     WyOming  56043.0   2358    44.0
2502831  2022!%05!%13      Weston     WyOming  56045.0   1588    18.0

[2445227 rows x 6 columns]

Luego si la columna Fips del data frame de los casos diarios es nula, entonces se pueden dar 2 casos:

El primero corresponde a que en la fila del condado diga desconocido (County = 'unknown'), para dicho caso no podemos recuperar el valor del Fips, por lo tanto la tupla también debe ser eliminada.

In [23]:
df_daily.loc[(df_daily['fips'].isnull()) & (df_daily['county'] == 'Unknown'), 'county'] = nan #cambiamos los valores de County de Unknown a NaN
df_daily = df_daily.dropna(subset=['county']) #eliminamos los nulos en County

El segundo caso corresponde a cuanto el Fips es nulo, pero conocemos el Condado y el Estado, para dicho caso podemos obtenter el Fips desde google y reemplazar el valor nulo por su correspondiente Fips, salvando así la información.

Si analizamos los datos nos daremos cuenta que solo sucede en 1 caso: Condado: New York City, Estado: New York, cuyo Fips es 36061.

In [24]:
df_daily.loc[(df_daily['fips'].isnull()) & (df_daily['county'] == 'New York CIty'), 'fips'] = 36061

Hay otros 2 casos, que  correspondían a los condados de Kansas City y Joplin, ambos en el estado de Missouri, sin embargo en la realidad estos nos son condados, sino que ciudades que abarcan varios condados que si se encuentran en el data frame, es por ello que pensamos que puede tratarse de datos duplicados. Decidimos eliminarlos.

In [25]:
df_daily.loc[(df_daily['fips'].isnull()) & (df_daily['county'] == 'Kansas CIty'), 'county'] = nan #volvemos el condado como nulo
df_daily.loc[(df_daily['fips'].isnull()) & (df_daily['county'] == 'JoplIn'), 'county'] = nan
df_daily = df_daily.dropna(subset=['county']) #eliminamos los nulos en County

In [26]:
df_daily.isnull().sum() #verificamos que ya no quedan valores nulos

date      0
county    0
state     0
fips      0
cases     0
deaths    0
dtype: int64

### Formateo de Datos

Notamos que los datos no están correctamente escritos, esto se puede solucionar con la funcion .title() que nos permite poner el texto en Camel Case (Primera letra con mayúscula y el resto con minúscula).

In [27]:
df_counties["county"] = df_counties["county"].str.lower()
df_counties["state"] = df_counties["state"].str.lower()
df_counties

fips               county        state state_code    male  female  \
0      1001       autauga county      alabama         AL   26874   28326   
1      1003       baldwin county      alabama         AL  101188  106919   
2      1005       barbour county      alabama         AL   13697   12085   
3      1007          bibb county      alabama         AL   12152   10375   
4      1009        blount county      alabama         AL   28434   29211   
...     ...                  ...          ...        ...     ...     ...   
3215  72145  vega baja municipio  puerto rico         PR   25580   27791   
3216  72147    vieques municipio  puerto rico         PR    4332    4439   
3217  72149   villalba municipio  puerto rico         PR   11169   11824   
3218  72151    yabucoa municipio  puerto rico         PR   16541   17608   
3219  72153      yauco municipio  puerto rico         PR   17475   18964   

      median_age        lat       long  
0           37.8  32.534923 -86.642730  
1           42.8  30.727479 -87.722564  
2           39.9  31.869581 -85.393210  
3           39.9  32.998628 -87.126475  
4           40.8  33.980869 -86.567380  
...          ...        ...        ...  
3215        40.7  18.428461 -66.397926  
3216        43.6  18.122662 -65.439095  
3217        38.8  18.128155 -66.472816  
3218        42.5  18.070468 -65.896311  
3219        43.0  18.079728 -66.858276  

[3220 rows x 9 columns]

In [28]:
df_states['county'] = df_states['county'].str.lower()
df_states['state'] = df_states['state'].str.lower()
df_states

county        state state_code  population
0          autauga county      alabama         AL       55200
1          baldwin county      alabama         AL      208107
2          barbour county      alabama         AL       25782
3             bibb county      alabama         AL       22527
4           blount county      alabama         AL       57645
...                   ...          ...        ...         ...
3215  vega baja municipio  puerto rico         PR       53371
3216    vieques municipio  puerto rico         PR        8771
3217   villalba municipio  puerto rico         PR       22993
3218    yabucoa municipio  puerto rico         PR       34149
3219      yauco municipio  puerto rico         PR       36439

[3220 rows x 4 columns]

In [29]:
df_daily['county'] = df_daily['county'].str.lower()
df_daily['state'] = df_daily['state'].str.lower()
df_daily

date      county       state     fips  cases  deaths
0        2020!%01!%21   snohomish  washington  53061.0      1     0.0
1        2020!%01!%22   snohomish  washington  53061.0      1     0.0
2        2020!%01!%23   snohomish  washington  53061.0      1     0.0
3        2020!%01!%24        cook    illinois  17031.0      1     0.0
4        2020!%01!%24   snohomish  washington  53061.0      1     0.0
...               ...         ...         ...      ...    ...     ...
2502827  2022!%05!%13  sweetwater     wyoming  56037.0  11088   126.0
2502828  2022!%05!%13       teton     wyoming  56039.0  10074    16.0
2502829  2022!%05!%13       uinta     wyoming  56041.0   5643    39.0
2502830  2022!%05!%13    washakie     wyoming  56043.0   2358    44.0
2502831  2022!%05!%13      weston     wyoming  56045.0   1588    18.0

[2422353 rows x 6 columns]

Notamos que para el data frame de los casos diarios, también convendría cambiar los Fips y las Muertes de formato, desde float a int.

In [30]:
df_daily = df_daily.astype({'fips': 'int64', 'deaths': 'int64'})
df_daily.dtypes

date      object
county    object
state     object
fips       int64
cases      int64
deaths     int64
dtype: object

In [31]:
df_daily['date'] = df_daily['date'].str.replace('!%', '-')
df_daily

date      county       state   fips  cases  deaths
0        2020-01-21   snohomish  washington  53061      1       0
1        2020-01-22   snohomish  washington  53061      1       0
2        2020-01-23   snohomish  washington  53061      1       0
3        2020-01-24        cook    illinois  17031      1       0
4        2020-01-24   snohomish  washington  53061      1       0
...             ...         ...         ...    ...    ...     ...
2502827  2022-05-13  sweetwater     wyoming  56037  11088     126
2502828  2022-05-13       teton     wyoming  56039  10074      16
2502829  2022-05-13       uinta     wyoming  56041   5643      39
2502830  2022-05-13    washakie     wyoming  56043   2358      44
2502831  2022-05-13      weston     wyoming  56045   1588      18

[2422353 rows x 6 columns]

Finalmente podemos cambiar la fecha de formato, desde object a date.

In [32]:
df_daily['date'] = pd.to_datetime(df_daily['date'])
df_daily.dtypes

date      datetime64[ns]
county            object
state             object
fips               int64
cases              int64
deaths             int64
dtype: object

## 3. Load

A continuación cargamos los datos en los data frames en tablas de sql en una base de datos de PostgreSQL

Primero seteamos correctamente los indices de cada data frame

In [33]:
df_counties.set_index('fips', inplace = True)
df_counties

county        state state_code    male  female  \
fips                                                                 
1001        autauga county      alabama         AL   26874   28326   
1003        baldwin county      alabama         AL  101188  106919   
1005        barbour county      alabama         AL   13697   12085   
1007           bibb county      alabama         AL   12152   10375   
1009         blount county      alabama         AL   28434   29211   
...                    ...          ...        ...     ...     ...   
72145  vega baja municipio  puerto rico         PR   25580   27791   
72147    vieques municipio  puerto rico         PR    4332    4439   
72149   villalba municipio  puerto rico         PR   11169   11824   
72151    yabucoa municipio  puerto rico         PR   16541   17608   
72153      yauco municipio  puerto rico         PR   17475   18964   

       median_age        lat       long  
fips                                     
1001         37.8  32.534923 -86.642730  
1003         42.8  30.727479 -87.722564  
1005         39.9  31.869581 -85.393210  
1007         39.9  32.998628 -87.126475  
1009         40.8  33.980869 -86.567380  
...           ...        ...        ...  
72145        40.7  18.428461 -66.397926  
72147        43.6  18.122662 -65.439095  
72149        38.8  18.128155 -66.472816  
72151        42.5  18.070468 -65.896311  
72153        43.0  18.079728 -66.858276  

[3220 rows x 8 columns]

In [34]:
df_states.set_index('county', inplace = True)
df_states

state state_code  population
county                                                 
autauga county           alabama         AL       55200
baldwin county           alabama         AL      208107
barbour county           alabama         AL       25782
bibb county              alabama         AL       22527
blount county            alabama         AL       57645
...                          ...        ...         ...
vega baja municipio  puerto rico         PR       53371
vieques municipio    puerto rico         PR        8771
villalba municipio   puerto rico         PR       22993
yabucoa municipio    puerto rico         PR       34149
yauco municipio      puerto rico         PR       36439

[3220 rows x 3 columns]

In [35]:
df_daily.set_index('date', inplace = True)
df_daily

county       state   fips  cases  deaths
date                                                    
2020-01-21   snohomish  washington  53061      1       0
2020-01-22   snohomish  washington  53061      1       0
2020-01-23   snohomish  washington  53061      1       0
2020-01-24        cook    illinois  17031      1       0
2020-01-24   snohomish  washington  53061      1       0
...                ...         ...    ...    ...     ...
2022-05-13  sweetwater     wyoming  56037  11088     126
2022-05-13       teton     wyoming  56039  10074      16
2022-05-13       uinta     wyoming  56041   5643      39
2022-05-13    washakie     wyoming  56043   2358      44
2022-05-13      weston     wyoming  56045   1588      18

[2422353 rows x 5 columns]

Luego procedemos a establecer la conexión con la base de datos.

In [36]:
engine = sqlalchemy.create_engine('postgresql://postgres:admin@localhost:5432/covidUsa')

Procedemos a cargar los datos a sus respectivas tablas

In [37]:
df_counties.to_sql('condados', engine, if_exists='append')

220

In [40]:
df_daily.to_sql('diarios', engine, if_exists = 'append')

353

In [39]:
df_states.to_sql('estados', engine, if_exists = 'append')

220